jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentCopier.java

```diff
@@ -46,43 +47,43 @@ public InvocationType getInvocationType()
     @Override
     protected Action process() throws Throwable
     {
+        if (current != null)
+            current.release();
```

Comment:
I think this is better done in an override of `succeeded()`, which BTW would not require the null check.

Reply:
@sbordet, as I keep saying (in other PRs), overriding succeeded and messing with fields there is an anti-pattern.  ICB enforces the concurrency that the process method is only ever called by a single thread.

A field that is changed in succeeded may be altered by the a call made from process, it may be done in the same thread, or a different one.  There is no memory boundary and the behaviour will be undefined.

The whole point of ICB is to serialize events into the process method, which can determine a specific ordering of them.   Recycling buffers that might be in use in another method is just not good practise, even if we "know" that the process method doesn't use them after scheduling an action.